### PR TITLE
[KEYCLOAK-8984] Warning in Gatekeeper on invalid state

### PR DESCRIPTION
**Pull request by abstractj**
Co-authored-by: Frode Ingebrigtsen <frode.ingebrigtsen@tellu.no>

### CODE REVIEW

File: handlers.go [hunk 1]
```diff
@@ -202,23 +202,25 @@ func (r *oauthProxy) oauthCallbackHandler(w http.ResponseWriter, req *http.Reque
 	}
 
 	// step: decode the state variable
-	state := "/"
+	redirectURI := "/"
 	if req.URL.Query().Get("state") != "" {
-		decoded, err := base64.StdEncoding.DecodeString(req.URL.Query().Get("state"))
-		if err != nil {
-			r.log.Warn("unable to decode the state parameter",
+		stateCookie, err := req.Cookie("OAuth_Token_Request_State")
```
[the remark this answers is not in the document]
Comment (abstractj) [reply handlers.go]: @FrodeI thank you! Good catch. I updated the code, based on your feedback.

**Comment by fredbi** [comment]
I have a question on this one. I had implemented the final landing page using encoded state, to accomodate the way it used to work.

I understand that now, setting the state for further checks by the client at the end of the sequence of redirection is no more possible.
Further, to control the final landing page, we must now use a special "request_uri" cookie...

Am I assuming correctly?

**Comment by fredbi** [comment]
Another remark: I cannot find the JIRA 8984 so it is difficult to grasp the background of this change.
The closest I could find is [KEYCLOAK-8856](https://issues.jboss.org/browse/KEYCLOAK-8856), but no mention there of the strategy adopted with this PR.

Could someone shed some light on this change?

**Comment by FrodeI** [comment]
Not sure I completely understand your setup. Did you use a modified Gatekeeper, or did you actually use the state from the protected application? (I assume it's from the cookie then).  

The behavior before this pull request was to pass the original URL in state (base64 encoded):
https://github.com/keycloak/keycloak-gatekeeper/pull/433
This was probably not a good solution since it basically removes the entropy from the state parameter. The pull request replaced the value of state with a uuid, but failed to replace state with another mechanism to communicate the original URL to the callback after authentication. This pull request fixes that by using another cookie (request_uri) to communicate the original URL. The protected application should hopefully not need to read this cookie, as Gatekeeper should return with a redirect to this URL as response to the authentication callback.

I filed issue 8984 because I noticed a warning message in the logs, and flagged in as an "Security Issue" since I was unsure if it affected validation of the state parameter. It didn't, but that's why the issue is not visible.

**Comment by fredbi** [comment]
I used the unmodified master from _just_ before this PR was merged, and passed an explicit state from the client (a JS application, using an iframe to manage the authN). Indeed, 

I understood that this was the way to retrieve a custom state after login (this allowed me to pass some parameter etc to close the iframe at the end of the handshake).

So now I had to adapt this setting and replace the encoded state query param by a cookie. This works now, with my client setting the cookie when calling /oauth/authorize.

I am just a bit wary that we break such a functionality without much announcement. I did not inspect much a PR named "fix warning xxx" and only realized it broke stuff, so this took me by surprise.

**Comment by abstractj** [comment]
@fredbi everything on _master_ branch is considered under development. For stable releases, please download the binaries from keycloak.org. We don't provide any announcements for changes into the codebase, instead, we use Jiras and keycloak-dev to keep track of everything. 

Anyways, if there's something you believe we should do better, feel free to bring this up on the mailing list.

**Pull request merged**